### ISSUE TITLE
Exporting single cell analysis txt from many cores in TMA

### ISSUE THREAD

**Issue by hartlandj** [qupath/qupath#59]
When exporting intensity measurements from a TMA I have a txt file for each level of the hierarchy but it would be necessary for me to know the corresponding TMA for all single cell intensities. Currently the txt is only one list of all single cells - an extra column corresponding to all higher levels of the hierarchy would be very useful.
Thanks!
Hart

**Comment by Svidro** [comment]
If you are referring to the core, that should be part of the "Name" column, or first column, in the detections measurements list (my first entry shows up as A-12 - Negative).  If you want to apply the name of the TMA, you would have to do that when you import the other TMA patient data, I think.

**Comment by hartlandj** [comment]
That is true for the TMA core export and my Tissue annotation but not for the txt output for all cells.
I'm referring to the Detection measurements for every core = single cells. Under the Name column it only says PathCellObject.
I've completed the cell analysis again and this is still the case.
Any suggestions appreciated.

**Comment by Svidro** [comment]
Hmm, interesting.  We are getting different behavior or choosing a different step order.  I am not sure I could get it to do that.  Are you using version 0.1.2?  And what OS (I don't think it should matter, but maybe Peter will know better)?

After running the dearrayer then selecting a core/cores to run cell detection, I see the following for cell by cell data.
![tma names](https://cloud.githubusercontent.com/assets/23145209/24367053/12f1cd4e-12d0-11e7-8fc0-dabcda42d2cd.JPG)

**Comment by hartlandj** [comment]
I think I'm doing the same...but end up with this:
![image](https://cloud.githubusercontent.com/assets/9638728/24367505/2fde4066-131d-11e7-840f-30c0611e64b5.png)

**Comment by petebankhead** [comment]
Ah... I was logging on to give the same answer as above, but @Svidro beat me to it.

But I see that the 'Name' column doesn't contain the TMA core if the cells are inside an annotation rather than directly inside the TMA core - which is the source of the problem, right?

I agree it would probably be useful to have the core always shown in the table, but in the absence of that you could use a simple script to repurpose the 'name' property for all your cells - setting it explicitly to be the same name as the core containing the cells.  That way it still appears in the table.

@Svidro already gave some of the code to do that [here](https://github.com/qupath/qupath/issues/57#issuecomment-289249459).
In your case, you might need to replace ```getParent()``` with ```getParent().getParent()```... to get the name of the object two levels up in the hierarchy rather than one - but the idea is the same.

In the event that you might have annotations inside annotations, then doing it that way would get a bit complicated since sometimes you might need to go up three levels... or four levels.
In that case, a simpler script would involve looping through all the TMA cores, and naming the cells inside them.

Does that sound like it would solve the problem, or have I misunderstood?

**Comment by hartlandj** [comment]
Sounds like that could work. I will play with the script and will also try running the cell analysis only one level down. It appears that Qupath currently applies the TMA name one level down but maybe not two.
Within my TMA cores, I run the tissue identification and then run the cell and membrane analysis within the annotated areas. Under Name the core identifier is present for the annotated tissue but not for the cells inside the annotated tissue.
Thanks, I'll get back to you tomorrow.
H

**Comment by Svidro** [comment]
Ah, I was just thinking that, but stuck in lab meeting!
Yes, I tend to calculate my tissue areas after the fact in R rather than using the simple tissue detection, so when I saw that line in your script, I figured that might be what was happening.  Good luck!

**Comment by hartlandj** [comment]
Yes - If I run the cell detection within each core (instead of within the annotation) then the core name is in the txt export. Thanks.
There are a couple of reasons that I would want to know which single cells are within my annotations, so the hierarchy export would be a useful enhancement in the future. 
Overall, great work!

**Comment by hartlandj** [comment]
A followup question:
Do you a have suggested package that you use for image analysis in R (alone or with QuPath)? For calculating tissue areas etc.
Thanks!

**Comment by petebankhead** [comment]
I'm not sure what you mean by 'hierarchy export'?
I think it would probably be good to have a column in the detection table giving the name of the TMA core containing the detection.  A column to add the immediate 'parent' object could also be helpful.  I suspect that trying to put more information concerning the hierarchy into the table could be confusing (since some detections might be at lower levels in the hierarchy than others), but I'll give it some thought.

I suspect @Svidro is much more familiar with R than I am, but if you are happy with R then I would suggest trying out some scripting in QuPath with Groovy.  There is a bit of documentation in the Wiki (starting [here](https://github.com/qupath/qupath/wiki/Writing-custom-scripts)).  It may take a little while to become familiar with QuPath's data structures, but I think Groovy has quite a nice syntax and [IntelliJ](https://github.com/qupath/qupath/wiki/Advanced-scripting-with-IntelliJ) can help considerably.

With even a little bit of scripting, you are no longer limited by the tables QuPath gives, but you can export your data any way you like... for use in R or elsewhere.

**Comment by Svidro** [comment]
Oops, I somehow missed the followup question.  I don't use any particular packages just to calculate areas, it's pretty much just a use of the base sum function.  I'm certain there are more elegant ways to do this using functions, but I only have basic programming experience so for loops it is!
```
path = "Your file path here"
setwd(path)
outFile <-"test areas.csv"

#Replace .txt with whatever identifier will pick up all of the files you want to analyze. Detections or Annotations are common choices
file.names <- dir(path,pattern = ".txt")

#an empty frame to place data into
Areas <- data.frame()

#simple for loop to read each file and keep a sum of the cell areas.
for(i in 1:length(file.names)){
data.raw <- read_delim(file.names[i],"\t", escape_double = FALSE, trim_ws = TRUE)

#place the file names in the first column
Sample = tools::file_path_sans_ext(file.names[i])
Areas[i,1]<-Sample

#Sum of cell areas here.  Add extra lines for mean of intensity/OD, etc
Areas[i,2]<-sum(data.raw$"Cell: Area")


}
Areas$"Area mm^2"<- Areas$V2/1000000

#set row names to F if you don't want a numbered list as the first column
write.csv(Areas, outFile, row.names=T)
```

**Comment by petebankhead** [comment]
The TMA core is now included in the detection table in QuPath v0.2.0-m3 (and likely earlier milestones...).
